### ISSUE TITLE
Naming conventions: <typeclass>Instances vs. <datatype>Instances

### ISSUE THREAD

**Issue by TomasMikula** [scalaz/scalaz#1533]
The meaning of "Instances" differs between `MonadInstances` and `DisjunctionInstances`:
 - `MonadInstances`: instances of *the* `Monad` typeclass for *various* data types;
 - `DisjunctionInstances`: instances of *various* typeclasses for *the* `Disjunction` data type.

I think it would clarify things, especially for beginners, if we could adopt a different naming convention for one of these meanings. Seeing `FooInstances`, we are currently not able to infer whether `Foo` is (in the role of) a typeclass or a datatype.

**Comment by NeQuissimus** [comment]
I would go even further...

Include companion objects, `Functions`, `Syntax` and `Module` and let's properly define what they include and mean. Are they even all needed?

I, for one, can't accurately figure out what they mean right now (so maybe I just need educating?) :)

**Comment by TomasMikula** [comment]
And `Class` and `Ops`... I agree that this should be clarified and part of the design guide.

I will just briefly comment on `Module`, since I am the one to blame for that one. We want to present some data types as abstract (or opaque; without exposing their representation), and provide functions to work with that abstract type. For example, we have

```scala
type Maybe[A] = Option[A]
```

but
 1. we want to hide the fact that `Maybe` is an alias for `Option`; and
 2. we can't have top-level type aliases, anyway.

So we put such abstract type inside a trait, called a module:

```scala
trait MaybeModule {
  type Maybe[A]

  // functions to work with Maybe
}
```

So a module declares an abstract type and functions to work with it.

In a package object of the containing package, we then have an implementation of the module:

```scala
package object data {
  val Maybe: MaybeModule = ???
  type Maybe[A] = Maybe.Maybe[A]
}
```

**Comment by jdegoes** [comment]
Agreed. All this should be part of the design guide. Same for Class, Ops, Syntax,e tc.

**Comment by TomasMikula** [comment]
Back to the point of this issue, just throwing in some ideas:

Rename `DisjunctionInstances` to `DisjunctionClasses`?

Alternatively, rename `MonadInstances` to `InstancesOfMonad` or `Monads`?

**Comment by jdegoes** [comment]
>  Rename DisjunctionInstances to DisjunctionClasses?

👍 This seems better to me. Or even `DisjunctionTypeclasses` if we want to go really verbose.

**Comment by NeQuissimus** [comment]
Things like `MonadInstances` only exist because they provide instances for types we do not control, correct? (i.e. the Scala standard library)
Why not get rid of all the `Instances` in `/typeclass/` and make a `/std/` folder, where we can instead split things in a way consistent with how we provide instances in `/data/`? `ListInstances` and `Function1Instances` etc.
I would stick with naming everything `[...]Instances` since it is familiar. But I would group everything by the target type, not the typeclass.

**Comment by TomasMikula** [comment]
That might work, too, but would require an additional `import`.

**Comment by jdegoes** [comment]
Technically those are orphan instances.

I think all instances for standard types must be defined in the companion object of the associated type classes. No additional import must be necessary to import them.

**Comment by NeQuissimus** [comment]
An extra import?
Can you not just throw them into the Prelude the same way it is done now?

**Comment by hrhino** [comment]
Now that the `std` module exists, is this still an issue?

**Comment by TomasMikula** [comment]
If we have eliminated all `<typeclass>Instances` occurrences, then this is no longer an issue.
